File: EngineKit/EngineKit/Math/Vector.m


#import "Vector.h"

#import "ObjectiveSugar.h"

#import "NSString+Extension.h"
#import "NSNumber+Extension.h"
#import "NSArray+Extension.h"
#import "NSDictionary+Extension.h"


@interface Vector ()
@property (nonatomic) SCNVector3 vector;
@end


unsigned long mutationsCounter;


@implementation Vector


//------------------------------------------------------------------------------
#pragma mark Getting the Vector's information
//------------------------------------------------------------------------------

// unavailable
- (NSNumber *)objectAtIndexedSubscript:(NSUInteger)index {
    switch (index) {
    case 0:
        return @(self.x);
        break;
    case 1:
        return @(self.y);
        break;
    case 2:
        return @(self.z);
        break;
    default:
        return nil;
        break;
    }

    return nil;
}

- (SCNVector3)toSCNVector3 {
    return self.vector;
}

- (SCNVector4)toSCNVector4 {
    return SCNVector4Make(self.x, self.y, self.z, 1.0);
}

- (NSValue *)toNSValue {
    return [NSValue valueWithSCNVector3:self.vector];
}

- (NSArray *)toArray {
    return @[ @(self.x), @(self.y), @(self.z) ];
}

- (BOOL)notZero {
    return !(self.x == 0 && self.y == 0 && self.z == 0);
}

//------------------------------------------------------------------------------
#pragma mark - Common Vector constants
//------------------------------------------------------------------------------

+ (instancetype)origin {
    static Vector *origin;

    if (!origin)
        origin = [[[self class] alloc] initWithX:0 Y:0 Z:0];

    return origin;
}


//------------------------------------------------------------------------------
#pragma mark - Creating Vector objects
//------------------------------------------------------------------------------

- (instancetype)init {
    self = [self initWithX:0 Y:0 Z:0];
    return self;
}

- (instancetype)initWithX:(CGFloat)x Y:(CGFloat)y Z:(CGFloat)z {
    if (self = [super init]) {
        self.vector = SCNVector3Make(x, y, z);
    }
    return self;
}

- (instancetype)initWithUniformNumbers:(CGFloat)x {
    self = [self initWithX:x Y:x Z:x];
    return self;
}

- (instancetype)initWithCGPoint:(CGPoint)newValue {
    self = [self initWithX:newValue.x Y:-newValue.y Z:0];
    return self;
}

- (instancetype)initWithCIVector:(CIVector *)newValue {
    self = [self initWithX:newValue.X Y:newValue.Y Z:newValue.Z];
    return self;
}

- (instancetype)initWithSCNVector3:(SCNVector3)newValue {
    self = [self initWithX:newValue.x Y:newValue.y Z:newValue.z];
    return self;
}

- (instancetype)initWithSCNVector4:(SCNVector4)newValue {
    self = [self initWithX:newValue.x Y:newValue.y Z:newValue.z];

    return self;
}

- (instancetype)initWithArray:(NSArray *)array {

    SCNVector3 vector;

    vector.x = [array floatAtIndex:0];
    vector.y = [array floatAtIndex:1];
    vector.z = [array floatAtIndex:2];

    self = [self initWithSCNVector3:vector];

    return self;
}

- (instancetype)initWithDictionary:(NSDictionary *)dictionary {
    SCNVector3 vector;

    vector.x = [dictionary floatForStringKey:@"x"];
    if (vector.x == 0.0)
        vector.x = [dictionary floatForStringKey:@"0"];

    vector.y = [dictionary floatForStringKey:@"y"];
    if (vector.y == 0.0)
        vector.y = [dictionary floatForStringKey:@"1"];

    vector.z = [dictionary floatForStringKey:@"z"];
    if (vector.z == 0.0)
        vector.z = [dictionary floatForStringKey:@"2"];

    self = [self initWithSCNVector3:vector];

    return self;
}

- (instancetype)initWithString:(NSString *)string {
    NSScanner *scanner = [NSScanner scannerWithString:string];
    NSCharacterSet *numbers =
        [NSCharacterSet characterSetWithCharactersInString:@"-0123456789."];

    NSMutableArray *array = [NSMutableArray array];

    BOOL done = NO;
    int contents = 0;

    until(done || contents == 3) {
        NSString *number;

        [scanner scanUpToCharactersFromSet:numbers intoString:NULL];

        done = ![scanner scanCharactersFromSet:numbers intoString:&number];

        if (number) {
            [array push:number.numberValue];
            contents++;
        }
    }

    self = [self initWithArray:array];

    return self;
}


- (instancetype)initWithVector:(Vector *)vector {
    self = [self initWithSCNVector3:[vector toSCNVector3]];

    return self;
}

- (instancetype)initWithObject:(id)object {
    if ([object isKindOfClass:[NSNumber class]]) {
        self = [self initWithUniformNumbers:((NSNumber *)object).doubleValue];
    } else if ([object isKindOfClass:[NSArray class]]) {
        self = [self initWithArray:object];
    } else if ([object isKindOfClass:[NSString class]]) {
        self = [self initWithString:(NSString *)object];
    } else if ([object isKindOfClass:[NSDictionary class]]) {
        self = [self initWithDictionary:(NSDictionary *)object];
    } else if ([object isKindOfClass:[Vector class]]) {
        self = [self initWithVector:object];
    } else {
        assert(false);
        return nil;
    }
    return self;
}


//------------------------------------------------------------------------------
#pragma mark - Comparing objects
//------------------------------------------------------------------------------

- (BOOL)isEqual:(id)object {
    return [self isEqualToVector:[[Vector alloc] initWithObject:object].vector];
}

- (BOOL)isEqualToVector:(SCNVector3)vector {
    return SCNVector3EqualToVector3(self.vector, vector);
}


//------------------------------------------------------------------------------
#pragma mark - Operations with Vectors
//------------------------------------------------------------------------------

- (Vector *)times:(CGFloat)scalar {

    return [[[self class] alloc] initWithX:self.x * scalar
                                   Y:self.y * scalar
                                   Z:self.z * scalar];
}

- (Vector *)over:(CGFloat)scalar {
    scalar = scalar ?: 1;
    return [[[self class] alloc] initWithX:self.x / scalar
                                   Y:self.y / scalar
                                   Z:self.z / scalar];
}

- (Vector *)plus:(Vector *)vector {
    return [[[self class] alloc] initWithX:self.x + vector.x
                                   Y:self.y + vector.y
                                   Z:self.z + vector.z];
}

- (Vector *)minus:(Vector *)vector {
    return [[[self class] alloc] initWithX:self.x - vector.x
                                   Y:self.y - vector.y
                                   Z:self.z - vector.z];
}

- (Vector *)opposite {
    return [[[self class] alloc] initWithX:-self.x Y:-self.y Z:-self.z];
}

- (CGFloat)dot:(id)object {
    Vector *vector = [[Vector alloc] initWithObject:object];
    return self.x * vector.x + self.y * vector.y + self.z * vector.z;
}

- (CGFloat)normSquared {
    return [self dot:self];
}

- (CGFloat)norm {
    return sqrt([self normSquared]);
}

- (Vector *)normalize {
    return [self over:[self norm]];
}

- (Vector *)translate:(id)object {
    Vector *vector = [[[self class] alloc] initWithObject:object];
    return [vector plus:self];
}

- (Vector *)scale:(Vector *)vector {
    return [[[self class] alloc] initWithX:self.x * vector.x
                                   Y:self.y * vector.y
                                   Z:self.z * vector.z];
}

- (SCNMatrix4)translateMatrix:(SCNMatrix4)matrix {
    return SCNMatrix4Translate(matrix, self.x, self.y, self.z);
}

- (SCNMatrix4)scaleMatrix:(SCNMatrix4)matrix {
    return SCNMatrix4Scale(matrix, self.x, self.y, self.z);
}

//------------------------------------------------------------------------------
#pragma mark - Overriding
//------------------------------------------------------------------------------

- (NSString *)description {
    return [NSString stringWithFormat:@"%@ %@ %@",
            [NSString stringFromFloat:self.vector.x],
            [NSString stringFromFloat:self.vector.y],
            [NSString stringFromFloat:self.vector.z]];

//    return [NSString stringWithFormat:@"(x = %lf, y = %lf, z = %lf)",
//                                      self.vector.x,
//                                      self.vector.y,
//                                      self.vector.z];
}

- (Vector *)setNewX:(CGFloat)x {
    return [[[self class] alloc] initWithX:x Y:self.vector.y Z:self.vector.z];
}

- (CGFloat)x {
    return self.vector.x;
}

- (Vector *)setNewY:(CGFloat)y {
    return [[[self class] alloc] initWithX:self.vector.x Y:y Z:self.vector.z];
}

- (CGFloat)y {
    return self.vector.y;
}

- (Vector *)setNewZ:(CGFloat)z {
    return [[[self class] alloc] initWithX:self.vector.x Y:self.vector.y Z:z];
}

- (CGFloat)z {
    return self.vector.z;
}

@end

File: EngineKit/EngineKit/Math/Vector.swift


import Foundation
import SceneKit


let originVector = Vector(x:0, y:0, z:0)


public class Vector: DebugPrintable {

    let vector: SCNVector3

    public var debugDescription: String {
        get {
            return String(format: "(x = %lf, y = %lf, z = %lf)",
                          arguments: [vector.x, vector.y, vector.z])
        }
    }

    public var x: Float { get { return vector.x } }
    public var y: Float { get { return vector.y } }
    public var z: Float { get { return vector.z } }

    init() {
        vector = SCNVector3Zero
    }

//------------------------------------------------------------------------------
// MARK: - Getting the Vector's information
//------------------------------------------------------------------------------

    public func notZero() -> Bool {
        return !(x == 0 && y == 0 && z == 0)
    }

    public func toArray() -> [Float] {
        return [x, y, z]
    }

    public func toSCNVector3() -> SCNVector3 {
        return vector
    }

    public func toSCNVector4() -> SCNVector4 {
        return SCNVector4Make(x, y, z, 0)
    }

//------------------------------------------------------------------------------
// MARK: - Creating Vector objects
//------------------------------------------------------------------------------

    public static func origin() -> Vector {
        return originVector
    }

    public init(x: Float, y: Float, z: Float) {
        vector = SCNVector3Make(x, y, z)
    }

    public convenience init(point: CGPoint) {
        self.init(x:Float(point.x), y:Float(-point.y), z:0.0)
    }

    public convenience init(number: Float) {
        self.init(x: number, y: number, z: number)
    }

    public convenience init(array: [Float]) {
        var x: Float = 0.0
        var y: Float = 0.0
        var z: Float = 0.0

        if (array.count > 0) {
            x = array[0]
        }
        if (array.count > 1) {
            y = array[1]
        }
        if (array.count > 2) {
            z = array[2]
        }

        self.init(x:x, y:y, z:z)
    }

    public convenience init(dictionary: [String : Float]) {
        var x: Float = 0.0
        var y: Float = 0.0
        var z: Float = 0.0

        if let result = dictionary["x"] {
            x = result
        } else if let result = dictionary["X"] {
            x = result
        } else if let result = dictionary["0"] {
            x = result
        }

        if let result = dictionary["y"] {
            y = result
        } else if let result = dictionary["Y"] {
            y = result
        } else if let result = dictionary["1"] {
            y = result
        }

        if let result = dictionary["z"] {
            z = result
        } else if let result = dictionary["Z"] {
            z = result
        } else if let result = dictionary["2"] {
            z = result
        }

        self.init(x:x, y:y, z:z)
    }

    public convenience init(string: String) {
        let scanner = NSScanner(string: string)
        let numbers = NSCharacterSet(charactersInString: "-0123456789.")
        var array: [Float] = []

        var done = false
        var contents = 0

        while (!done && contents != 3) {
            var number: NSString? = ""

            scanner.scanUpToCharactersFromSet(numbers, intoString: nil)

            done = !(scanner.scanCharactersFromSet(numbers, intoString: &number))

            if let float = number?.floatValue {
                array.append(float)
                contents++
            }
        }

        self.init(array:array)
    }

    public convenience init(vector: Vector) {
        self.init(x:vector.x, y:vector.y, z:vector.z)
    }

    public convenience init(object: Any) {
        if let point = object as? CGPoint {
            self.init(point: point)
        }
        else if let float = object as? Float {
            self.init(number: float)
        }
        else if let array = object as? [Float] {
            self.init(array: array)
        }
        else if let dictionary = object as? [String: Float] {
            self.init(dictionary: dictionary)
        }
        else if let string = object as? String {
            self.init(string: string)
        }
        else {
            self.init(x:0, y:0, z:0)
        }
    }

//------------------------------------------------------------------------------
// MARK: - Common overrides
//------------------------------------------------------------------------------

    public func isEqual(object: Any) -> Bool {
        if let other = object as? Vector {
            return SCNVector3EqualToVector3(vector, other.vector)
        }
        return false
    }

    public func isApproximatelyEqual(object: Any) -> Bool {
        let margin: Float = 0.00001

        if let otherVector = object as? Vector {
            let other = otherVector.vector
            if (Float.abs(vector.x - other.x) > Float.abs((vector.x * margin) + margin)) {
                return false
            }
            if (Float.abs(vector.y - other.y) > Float.abs((vector.y * margin) + margin)) {
                return false
            }
            if (Float.abs(vector.z - other.z) > Float.abs((vector.z * margin) + margin)) {
                return false
            }
        }
        return true
    }
}

infix operator ==~ {associativity left precedence 130}
infix operator !=~ {associativity left precedence 130}

public func == (left: Vector, right:Vector) -> Bool {
    return left.isEqual(right)
}

public func != (left: Vector, right:Vector) -> Bool {
    return !left.isEqual(right)
}

public func ==~ (left: Vector, right:Vector) -> Bool {
    return left.isApproximatelyEqual(right)
}

public func !=~ (left: Vector, right:Vector) -> Bool {
    return !left.isApproximatelyEqual(right)
}
